torchaudio/models/wav2vec2/components.py

```diff
@@ -397,6 +401,28 @@ def forward(
 
         return x
 
+    def get_intermediate_outputs(
+            self,
+            x: Tensor,
+            indices: Optional[int] = None,
+            attention_mask: Optional[Tensor] = None,
+    ) -> List[Tensor]:
+        if indices is None:
+            indices = list(range(len(self.layers)))
+        assert indices is not None
+        if len(indices) == 0:
+            return []
+        ret: List[Tensor] = []
+
+        x = self._preprocess(x)
+        for i, layer in enumerate(self.layers):
```

Comment:
this way of iterating through layers does not ensure that the order of the resulting list of layers corresponds to the order of the indices, since things are being added to the return list in order of increasing layer index. ex/ `indices=[2,1]` would return `[layer1, layer2]` as opposed to `[layer2, layer1]`

with the current implementation, the check for `i in indices` should precede the layer masking computation to avoid unnecessary computation

Reply:
To achieve the purpose of correct ordering, we can improve the current impl by
```
for i in range(max(indices) + 1):
    layer = self.layers[i]
    x = layer(x, attention_mask)
    ret.append(x)
return ret[indices]
```

Reply:
Thanks, the current signature also allows the duplication of the layers.
I think it would be easier to resort to the `fairseq`'s signature where you provide the index of the last layer you need and return all the results from the previous layers. That way we do not need to worry about the order and duplication. We only need to worry if the index user provided is within the number of existing layers.

Let's continue this on  #1776